### PR TITLE
Add API to retrieve package files and directories, in a cross-platform way.

### LINKED ISSUE #1551: [Enhancement] Need for FileSystem.AppResourcesDirectory
## Summary
On _iOS_ and _macOS_ the call to **Directory.GetCurrentDirectory ()**; resolves to the **Resources** directory while on Android it just resolves to **/**.

I need all of them to resolve consistently to the **Resources** directory under _iOS_/_macOS_ and to the **Assets** folder under Android because I have a subdirectory in those locations, that vary from app to app, so I need to traverse/recurse through them to load all the required files on startup.

Now Xamarin.Essentials has **FileSystem.OpenAppPackageFileAsync**, which is great if you know the exact name of the file you want to load, but not if you want to traverse/recurse sub-directories.

## API Changes
I would like a `string FileSystem.AppResourcesDirectory` property that just resolves to **Resources** directory under _iOS_/_macOS_ and to the **Assets** folder under Android and the respective equivalents under Windows, Tizen and elsewhere.

## Intended Use Case
I need all of them to resolve consistently to the **Resources** directory under _iOS_/_macOS_ and to the **Assets** folder under Android because I have a subdirectory in those locations, that vary from app to app, so I need to traverse/recurse through them to load all the required files on startup.

If there is already a cross-platform way to do this, apologies for the waste of bandwidth :)

### PR DESCRIPTION
**Pull request by CartBlanche**
### Description of Change ###

Given a Directory name,  respectively return either the sub-directories or files within that directory

### Bugs Fixed ###

- Fixes #1551

### API Changes ###

Added: 

```csharp
public static partial class FileSystem {
    .
    .
    public string[] GetAppPackageDirectories(string path);
    public string[] GetAppPackageFiles(string path);
}
```

And associated cross-platform code.

### Behavioural Changes ###

None. New functionality.

### PR Checklist ###

- [x] Has tests (if omitted, state reason in description)
- [x] Has samples (if omitted, state reason in description)
- [x] Rebased on top of `main` at time of PR
- [x] Changes adhere to coding standard
- [ ] Updated documentation ([see walkthrough](https://github.com/xamarin/Essentials/wiki/Documenting-your-code-with-mdoc))

### Sample UI Changes ###
**Android:**
Android Screenshot

**iOS:**
iOS Screenshot

**Mac:**
macOS Screenshot

**Tizen:**
Tizen Screenshot

**Windows.UWP:**
Windows Screenshot

### CODE REVIEW

**Comment by CartBlanche** [comment]
Adding @mattleibow 's original PR review for context - https://github.com/xamarin/Essentials/pull/1566#pullrequestreview-570131500

**Comment by CartBlanche** [comment]
@mattleibow, @jamesmontemagno with the upcoming **Xamarin.iOS 15.x**  release, the Sample.iOS will no longer compile due to warnings as errors, because the UIApplication.Main you are using uses the now obsolete nameof() call. You may want to grab this commit 6fb05ae18e3dcf35e66d8c0241a81f8bdd9549d5 at some point, which uses typeof() instead, assuming that is already available in Xamarin.iOS 14.x.

File: Xamarin.Essentials/FileSystem/FileSystem.ios.tvos.watchos.macos.cs [hunk 1]
```diff
@@ -38,6 +41,34 @@ static string GetDirectory(NSSearchPathDirectory directory)
             }
             return dirs[0];
         }
+
+        static string[] PlatformGetAppPackageDirectories(string path)
+        {
+            if (string.IsNullOrEmpty(path))
+                throw new ArgumentNullException(nameof(path));
+
+            if (path.StartsWith("/"))
```
Comment (CartBlanche) [inline Xamarin.Essentials/FileSystem/FileSystem.ios.tvos.watchos.macos.cs:50]: This is not required on Android as it defaults to the root of the Asset files list. If either of you think that there's a nicer way to ensure root paths aren't used, let me know.

**Comment by CartBlanche** [comment]
I'm currently using this new API in 2 projects. A commercial language learning project (iOS, Android, Mac and Windows UWP) for a customer, as well as a new personal gaming project.

**Comment by CartBlanche** [comment]
@mattleibow, @jamesmontemagno I cannot see what the CI build error is for me to know why it's failing. Says I'm not authorised to click on "Details" to see anything. Any heads up greatly appreciated.

**Comment by jfversluis** [comment]
Hey @CartBlanche thank you so much for this PR and the time and effort on this. However, unfortunately at this time we're unable to add new things to Xamarin.Essentials. If this is still of interest to you, please open a discussion or issue on the .NET MAUI repo to see if we can add this to that.

Thanks!

**Comment by CartBlanche** [comment]
> Hey @CartBlanche thank you so much for this PR and the time and effort on this. However, unfortunately at this time we're unable to add new things to Xamarin.Essentials. If this is still of interest to you, please open a discussion or issue on the .NET MAUI repo to see if we can add this to that.
> 
> Thanks!

I somehow missed that this PR had been closed. So does MAUI have visual and non visual APIs that are separated or does one pull in the whole of MAUI to see if my proposed API already exists?

**Pull request closed without merging**